### PR TITLE
[Misc] Enhance attention selector

### PR DESCRIPTION
**Pull request by WoosukKwon**
This PR is to provide more information (such as block size and kv cache dtype) to attention backend selector so that it can be used to find the appropriate attention backend. Also, the PR moves `kv_cache_dtype` from `AttentionMetadata` to `Attention`.

This PR is a prerequisite for #3648

### CODE REVIEW

File: vllm/attention/layer.py [hunk 1]
```diff
@@ -29,10 +30,22 @@ def __init__(
         num_kv_heads: Optional[int] = None,
         alibi_slopes: Optional[List[float]] = None,
         sliding_window: Optional[int] = None,
+        cache_config: Optional[CacheConfig] = None,
```
Comment (rkooo567) [inline vllm/attention/layer.py:33]: QQ: when is this None? (should we just not allow None here? Since cache config is supposed to be created by default?)

File: vllm/attention/selector.py [hunk 2]
```diff
@@ -44,14 +54,22 @@ def get_attn_backend(dtype: torch.dtype) -> Type[AttentionBackend]:
...
+        logger.warning("Eager mode is enforced for the Flashinfer backend.")
         from vllm.attention.backends.flashinfer import FlashInferBackend
         return FlashInferBackend
     else:
         raise ValueError("Invalid attention backend.")
 
 
-def _which_attn_to_use(dtype: torch.dtype) -> _Backend:
+def _which_attn_to_use(
+    num_heads: int,
```
Comment (rkooo567) [inline vllm/attention/selector.py:65]: is this change necessary? (seems like most of args are not used?)

File: vllm/attention/layer.py [hunk 3]
```diff
@@ -29,10 +30,22 @@ def __init__(
...
-        impl_cls = self.backend.get_impl_cls()
+        if cache_config is not None:
+            kv_cache_dtype = cache_config.cache_dtype
+            block_size = cache_config.block_size
+        else:
+            kv_cache_dtype = "auto"
+            block_size = 16
+        if num_kv_heads is None:
+            num_kv_heads = num_heads
+        dtype = torch.get_default_dtype()
```
Comment (rkooo567) [inline vllm/attention/layer.py:44]: Can you comment default dtype is selected when a model is loaded?

**Review by rkooo567 — APPROVED** [review]
LGTM! One question regarding why we change the interface of `get_attn_backend`!

File: vllm/attention/selector.py (hunk 2 above)
Comment (WoosukKwon) [reply vllm/attention/selector.py:65]: Good question! It's actually for the PR #3648 and future PRs where we need to consider block sizes and KV cache dtypes in selecting the backend.

File: vllm/attention/layer.py (hunk 3 above)
Comment (WoosukKwon) [reply vllm/attention/layer.py:46]: Sure! Added.

File: vllm/attention/layer.py (hunk 1 above)
Comment (WoosukKwon) [reply vllm/attention/layer.py:33]: Good point. In most situations, `cache_config` isn't `None`. However, I wanted to provide the flexibility to initialize the model without `cache_config`, which can be particularly useful in niche scenarios such as testing the model loader. For instance, some tests in `test_tensorizer` only use the HF config to initialize the model, without setting up a `CacheConfig` or `ModelConfig`. Additionally, allowing `cache_config` to be optional helps maintain consistency with the HF model interface, where a model can be instantiated solely with the HF config. 
I think this adjustment makes the setup more versatile and aligns better with existing practices.
